### DISCUSSION TITLE
Support for Oxlint

### DISCUSSION THREAD

**Discussion by fantoine** [nrwl/nx#21439] [Feature Requests]
## Description

Allow the usage of [oxlint](https://oxc-project.github.io/docs/guide/usage/linter.html) in place of eslint.

## Motivation

Even if oxlint is still pretty young, it provides far better performances than eslint ([50 to 100 times faster according to them](https://oxc-project.github.io/blog/2023-12-12-announcing-oxlint.html#_50-100-times-faster-than-eslint)).

Since in large projects eslint can run for several minutes, this would significantly improve DX and CI integrations.

## Suggested Implementation

I suppose that adding a new `@nx/oxlint` executor should be the way to go.

**Comment by AgentEnder** [comment]
Hey @fantoine - this would be a good case for a community plugin to work with oxlint and test out the dev ex while the tool gains traction.

**Comment by yzhe554** [comment]
From my experience with `oxlint`, it's amazing but I would say it isn't close to complete.
It could be faster than NX cache as it uses rust :)

**Comment by pumano** [comment]
Today I test oxlint against my large codebase ( 90 libs and around 8k files ).

Default eslint / prettier (and other rules) pipeline is ~12 minutes without cache
oxlint (eslint + import) 378ms!!

That will be new default linter 100%. But we really don't need nx plugin because whole REPO linted for half of sec! I think maximum time will be few secs with support all needed plugins in future!

**Reply by pumano** [reply]
nx plugin I think a good idea when you want to lint specific library, not a whole repo.

**Comment by ild0tt0re** [comment]
Yes it is really blazing fast! We are waiting for type-aware support.

**Comment by hypeJunction** [comment]
FYI oxlint is now out of beta https://voidzero.dev/posts/announcing-oxlint-1-stable

**Comment by Stanzilla** [comment]
I think NX should have an official oxc plugin

**Comment by Nas3nmann** [comment]
I just released an community nx plugin for oxlint. Give it a shot here: https://github.com/Nas3nmann/nx-oxlint

**Comment by juristr** [comment]
Thanks @Nas3nmann for the community plugin. Great work.

@Stanzilla we're looking into something like that. Just some other things we need to clear first in our pipeline, but it is on my radar for sure! So stay tuned :)

**Reply by Stanzilla** [reply]
thanks! very curious to hear what the plans are regarding Vite+ tooling and NX since they are sort of competitors as a complete package but also all Vite+ parts usable standalone

**Comment by intellix** [comment]
I'm using the nx-oxlint plugin and doing a progressive migration:
```json
"lint.es": {
  "executor": "@nx/eslint:lint",
  "options": {
    "lintFilePatterns": ["libs/ui-atoms/**/*.{ts,tsx,js,jsx}"]
  }
},
"lint.ox": {
  "executor": "nx-oxlint:lint",
  "options": {
    "configFile": "./.oxlintrc.json"
  }
},
"lint": {
  "dependsOn": ["lint.ox", "lint.es"]
}
```

The difference for linting that lib is initially 200ms vs 3sec, but once I add the `@nx/enforce-module-boundaries` plugin, it jumps from 200ms to 1.7sec for that library.

I guess at some point in the future it would be nice if there was a Rust Plugin as opposed to JS Plugin to take advantage of the πerf benefits from oxide linting. 

Thanks for the library @Nas3nmann !

**Comment by the-ult** [comment]
With OXLint and OXfmt getting more and more tracktion. Wouldn't it still be a good idea to make OXLint a default plugin of nx as well? Next to eslint?

And the same for the module-boundaries etc. Could that be oxlint compatible / oxlint-first? (so we do not have to rely on the OXLint JS Plugin)

**Reply by Stanzilla** [reply]
https://github.com/nrwl/nx/pull/34838

**Comment by xiongemi** [comment]
there is an open pr in nx-labs: https://github.com/nrwl/nx-labs/pull/441

**Comment by xiongemi** [comment]
there is a new draft pr: https://github.com/nrwl/nx/pull/36491 for nx oxlint plugin

**Comment by ctjhoa** [comment]
For the record, oxlint and oxfmt support has landed in [Nx 23.2.0](https://github.com/nrwl/nx/releases/tag/23.2.0)
